src/threads.rs

```diff
@@ -0,0 +1,382 @@
+pub struct ThreadId(usize);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(id as usize)
+    }
+}
+
+impl From<u32> for ThreadId {
+    fn from(id: u32) -> Self {
+        Self(id as usize)
+    }
+}
+
+impl ThreadId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(u32::try_from(self.0).unwrap())
+    }
+}
+
+/// An identifier of a set of blocked threads.
+///
+/// Note: 0 is not a valid identifier.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockSetId(u32);
+
+impl From<u32> for BlockSetId {
+    fn from(id: u32) -> Self {
+        assert_ne!(id, 0, "0 is not a valid blockset id");
+        Self(id)
+    }
+}
+
+impl BlockSetId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(self.0)
+    }
+}
+
+/// The state of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ThreadState {
+    /// The thread is enabled and can be executed.
+    Enabled,
+    /// The thread tried to join the specified thread and is blocked until that
+    /// thread terminates.
+    BlockedOnJoin(ThreadId),
+    /// The thread is blocked and belongs to the given blockset..
+    Blocked(BlockSetId),
+    /// The thread has terminated its execution (we do not delete terminated
+    /// threads.)
+    Terminated,
+}
+
+/// A thread.
+pub struct Thread<'mir, 'tcx> {
+    state: ThreadState,
+    /// Name of the thread.
+    thread_name: Option<Vec<u8>>,
+    /// The virtual call stack.
+    stack: Vec<Frame<'mir, 'tcx, Tag, FrameData<'tcx>>>,
+    /// Is the thread detached?
+    ///
+    /// A thread is detached if its join handle was destroyed and no other
+    /// thread can join it.
+    detached: bool,
+}
+
+impl<'mir, 'tcx> Thread<'mir, 'tcx> {
+    /// Check if the thread terminated. If yes, change the state to terminated
+    /// and return `true`.
+    fn check_terminated(&mut self) -> bool {
+        if self.state == ThreadState::Enabled {
+            if self.stack.is_empty() {
+                self.state = ThreadState::Terminated;
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl<'mir, 'tcx> std::fmt::Debug for Thread<'mir, 'tcx> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.state)
+    }
+}
+
+impl<'mir, 'tcx> Default for Thread<'mir, 'tcx> {
+    fn default() -> Self {
+        Self { state: ThreadState::Enabled, thread_name: None, stack: Vec::new(), detached: false }
+    }
+}
+
+/// A set of threads.
+#[derive(Debug)]
+pub struct ThreadManager<'mir, 'tcx> {
+    /// Identifier of the currently active thread.
+    active_thread: ThreadId,
+    /// Threads used in the program.
+    ///
+    /// Note that this vector also contains terminated threads.
```

Comment:
This includes joined threads? Could we have memory leak issues here?

Reply:
> This includes joined threads?

Yes. It includes all threads ever created (this is a add-only vector).

> Could we have memory leak issues here?

Not sure. Could you elaborate what kind of memory leaks do you have in mind? As far as I understand, the `leak_report` function does not depend on the stack in any way, so this should not be a problem for it.

Reply:
I meant leaks on the host. :D If the program spawns and joins many threads, Miri will eat more and more RAM.

Why are dead threads still needed here? Could they be replace by some kind of "tombstone" that uses minimal amounts of memory?

Reply:
> I meant leaks on the host. :D If the program spawns and joins many threads, Miri will eat more and more RAM.

The struct currently takes 72 bytes + size of the stack. 72 bytes per thread does not seem like a problem to me and we could release the stack memory for each terminated thread with `thread.stack = Vec::new()` if you think that is necessary.

> Why are dead threads still needed here?

This design is mostly based on my experience that it is easier to debug programs that do not deallocate things. Also, currently, we are running deallocation of thread locals [after](https://github.com/rust-lang/miri/blob/03715907f46f79250904c799a59b87c26c2b0377/src/eval.rs#L210-L211) all threads have finished.

Reply:
> Also, currently, we are running deallocation of thread locals after all threads have finished.

That sounds like it would incorrectly miss UB though if other threads still access that memory. I think this should be fixed -- or at least marked as a FIXME in his PR with an issue opened that is flagged "misses UB".

Reply:
> I think this should be fixed -- or at least marked as a FIXME in his PR with an issue opened that is flagged "misses UB".

I have changed the behaviour to execute dtors on Linux and Mac immediately when a thread terminates.

Reply:
Thanks! Please also add a compile-fail test to ensure this catches UB properly.